### ISSUE TITLE
How to disable AWS Secrets per profile

### ISSUE THREAD

**Issue by spekdrum** [awspring/spring-cloud-aws#824]
Hi, since 3.0.0 version we can't disable secrets in local or test profiles easily. 

It seems this property has been removed:

`spring.cloud.aws.secretsmanager.enabled`

Now the only way I see to disable it is excluding some auto configuration classes.

```
  io.awspring.cloud.autoconfigure.core.AwsAutoConfiguration
  io.awspring.cloud.autoconfigure.core.CredentialsProviderAutoConfiguration
```

If we don't do that we are getting this error:
`Unable to load region from any of the providers in the chain...`

So we have to set this:
`spring.cloud.aws.region.static=eu-west-1`
and ensure we dont set it in production profile, for example.

Is this correct? Is there any cleaner way to disable Secrets Manager in the app settings so we can disable it in certain profiles?

thanks!

**Comment by poklakni** [comment]
same for parameter store

**Comment by eddumelendez** [comment]
This is being activated because there is spring.config.import: aws-secretsmanager:// sonewhere is the application.properties . Why declaring the property if it is not going to be used? I would like to understand

**Comment by ipalbeniz** [comment]
I think that the property `spring.config.import` cannot be modified or disabled in a profile specific application.properties, so the `spring.cloud.aws.secretsmanager.enabled` flag would be very helpful to do that

**Comment by poklakni** [comment]
one way it might be possible is like this

```yml
here:
  is:
    your:
      other:
        config: blabla

---
spring:
  config:
    activate:
      on-profile: "!yourdevprofile"
    import:
      - aws-parameterstore:/your/config
```

**Comment by spekdrum** [comment]
Hi @eddumelendez 
> This is being activated because there is spring.config.import: aws-secretsmanager:// sonewhere is the application.properties . Why declaring the property if it is not going to be used? I would like to understand

The error I mentioned occurs without any config import in app.properties

**Comment by poklakni** [comment]
@spekdrum you get this error because of software.amazon.awssdk:sts:2.20.72 dependency, not because of secrets manager

**Comment by spekdrum** [comment]
@poklakni I see, thanks. So this is not directly related with secrets, but the fact is that we need STS in order to access to aws.
Maybe the question is how to disable the STS initialization per profile.

**Comment by MatejNedic** [comment]
Hey @spekdrum , there is a way.

Firstly I strongly recommend setting region property via properties. It will create `AwsRegionProvider` bean that our spring cloud aws integrations will use when configuring a clients.

If you dont want sts to be configured in credentials provider simply do not metion any `spring.cloud.aws.credentials.sts` properties or `AWS_WEB_IDENTITY_TOKEN_FILE` or the javaproperty `aws.webIdentityTokenFile`.  Then sts won't be configured.

Hope it helps.

**Comment by spekdrum** [comment]
Hi @MatejNedic, The region setting seems to be mandatory if I add the sts dependency. When I add the dependency, even with an empty application.yml, the exception shows up. Thanks

**Comment by ipalbeniz** [comment]
In Spring Cloud Vault, you can disable importing Vault secrets by using two configuration properties:

- `spring.config.import=optional:vault://`
- `spring.cloud.vault.enabled=false`

Wouldn't it be better to use a common way to disable it, either using AWS Secrets Manager or Vault?

For more information, please refer to the following link: [Spring Cloud Vault Documentation - 4.2. Conditionally enable/disable Vault Configuration](https://docs.spring.io/spring-cloud-vault/docs/current/reference/html/#vault.configdata.location.optional)

**Comment by ipalbeniz** [comment]
Hey @MatejNedic,

I'll break it down for you. 

Our applications use 4 profiles:
- local: for the local environment where we don't use Config Server.
- dev: for the development environment where Config Server is needed.
- stage: for the staging environment where Config Server is needed.
- pro: for the production environment where Config Server is needed.


Previously, we only had **2 yml files** in the application repository:
- application.yml: the base configuration with Config Server enabled by default.
- application-local.yml: the local configuration with Config Server disabled.


But now, with the new `spring.config.import` property, we require **5 yml files** in the application repository:
- application.yml: the base configuration. Unfortunately, we can't configure the `spring.config.import` to point to the config server because this property **can't be overridden in other profile-specific yml files** (like application-local.yml).
- application-local.yml: the local configuration.
- application-dev.yml: the development configuration with the `spring.config.import` pointing to the config server.
- application-stage.yml: the staging configuration with the `spring.config.import` pointing to the config server.
- application-pro.yml: the production configuration with the `spring.config.import` pointing to the config server.


So, The real pain point is that the `spring.config.import` is not overridable. You can read the last comment in this issue, having 11 likes but ignored: https://github.com/spring-cloud/spring-cloud-config/issues/1877#issuecomment-953558794

**Comment by ipalbeniz** [comment]
Sorry, when I say 'Config Server,' it can be changed to 'Secrets Manager.'

**Comment by maciejwalkowiak** [comment]
I think it makes sense to have similar option to disable it as @ipalbeniz mentioned Spring Cloud Vault provides.

**Comment by poklakni** [comment]
> Hey @MatejNedic,
> 
> I'll break it down for you.
> 
> Our applications use 4 profiles:
> 
>     * local: for the local environment where we don't use Config Server.
> 
>     * dev: for the development environment where Config Server is needed.
> 
>     * stage: for the staging environment where Config Server is needed.
> 
>     * pro: for the production environment where Config Server is needed.
> 
> 
> Previously, we only had **2 yml files** in the application repository:
> 
>     * application.yml: the base configuration with Config Server enabled by default.
> 
>     * application-local.yml: the local configuration with Config Server disabled.
> 
> 
> But now, with the new `spring.config.import` property, we require **5 yml files** in the application repository:
> 
>     * application.yml: the base configuration. Unfortunately, we can't configure the `spring.config.import` to point to the config server because this property **can't be overridden in other profile-specific yml files** (like application-local.yml).
> 
>     * application-local.yml: the local configuration.
> 
>     * application-dev.yml: the development configuration with the `spring.config.import` pointing to the config server.
> 
>     * application-stage.yml: the staging configuration with the `spring.config.import` pointing to the config server.
> 
>     * application-pro.yml: the production configuration with the `spring.config.import` pointing to the config server.
> 
> 
> So, The real pain point is that the `spring.config.import` is not overridable. You can read the last comment in this issue, having 11 likes but ignored: [spring-cloud/spring-cloud-config#1877 (comment)](https://github.com/spring-cloud/spring-cloud-config/issues/1877#issuecomment-953558794)

@ipalbeniz I assume you are migrating from version 2.x.x to 3.x.x.
In 2.x.x the config loading was based on the prefix + application name (or explicitly configured name) + spring profile name.

Lets work with example `/secret/dummy-app_stage where `dummy-app` is you application name and `stage` is your profile name for stage env.

In v3.x.x you can achieve the same functionality even without creating additional yaml files.

you application.yml could look like this
```yaml
here:
  is:
    your:
      other:
        config: johndoe

---
spring:
  config:
    activate:
      on-profile: "!local" #spring.config.import won't be applied for local profile
    import:
      - aws-secretsmanager:/secret/application/
      - aws-secretsmanager:/secret/application_${ENV}/ #ENV is an environmental variable that would hold the same value as your profile, in this case 'stage'
      - aws-secretsmanager:/secret/dummy-app/
      - aws-secretsmanager:/secret/dummy-app_${ENV}
```

**Comment by ipalbeniz** [comment]
Thanks @poklakni,

I didn't want to use **multi-document yaml files** because it is not so obvious what properties it contains based on the file name. When a developer finds an `application.yml` file (without any profile in its name), he expects it to contain only common properties.

We can find other _workarounds_ but I think that it was so simple and elegant to configure before the new `spring.config.import` property mechanism :(

**Comment by poklakni** [comment]
@maciejwalkowiak if you decide `spring.cloud.aws.parameterstore.enabled` should be added I can work on it.

**Comment by pdumontel** [comment]
I am having the same problem, but my import for aws-secretsmanager is activated by being on kubernetes, i.e.:
```
spring:
  config:
    activate:
      on-platform: kubernetes
``` 
which shouldn't be activated in a unit test, should it?

**Comment by spekdrum** [comment]
Hi, this no longer happens with recent versions. Thanks!

**Comment by ielkhalloufi** [comment]
I encountered the same issue, the "enabled" option is not existing, and it is not possible to override a "Bean" for testing purposes. See issue for an ugly workaround: https://github.com/awspring/spring-cloud-aws/issues/1005

**Comment by maciejwalkowiak** [comment]
Fix: https://github.com/awspring/spring-cloud-aws/pull/1009

**Comment by ielkhalloufi** [comment]
@maciejwalkowiak can we expect the same fix also for the ParameterStore: [#1004](https://github.com/awspring/spring-cloud-aws/issues/1004)

**Comment by maciejwalkowiak** [comment]
@ielkhalloufi yes, once this one is reviewed and merged Parameter Store will follow
